### DISCUSSION TITLE
Syntax for associating pictures with steps

### DISCUSSION THREAD

**Discussion by yb66** [cooklang/spec#58] [Ideas]
Hi,

I mentioned this on the HN thread, I can see the steps being fiddly and easy to get wrong, and clashes to occur as it's common to have 1,2,3… added to file names. Instead of this:

```
Chicken French.cook
Chicken French.0.jpg
Chicken French.3.jpg
```

I suggest the style for measures is used, e.g.

```
Chicken French.cook
Chicken French.jpg{0}
Chicken French.jpg{3}
```

Once it's ascertained that it's a file then what's in the braces can be inferred as a step, or just make it follow the ingredients rule:

```
Chicken French.cook
Chicken French.jpg{0%step}
Chicken French.jpg{3%step}
```

or even:

```
Chicken French.cook
Chicken French.jpg{step 0}
Chicken French.jpg{step%3}
```

Any or all would seem an improvement, in my eyes. However, I understand they may have different challenges in implementation.

Just my thoughts. I'll be transferring all my recipes over to use the spec, it looks really good 👌 

Regards,
iain

**Comment by dubadub** [comment]
Hi Iain, thanks for bringing this up.

I like your idea to include information in a markup way, definitely it will isolate meta-information from the filename and prevent collisions. There's also another issue people mentioned about: setting multiple images for the same step or title. This can be handled with your approach too.

The only thing is that on Windows instead of using "magic byte" for understanding the file type (like macOS or Linux does), it uses file extension. After adding meta-information in the end it makes Windows to forget that it's a picture, so previews won't work. I'm surprised but filename [can include these symbols](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions), which is a good news.

**Comment by deathau** [comment]
After the name, but before the extension might be a good fit, though.
```
Chicken French.cook
Chicken French{0}.jpg
Chicken French{3}.jpg
```

**Comment by yb66** [comment]
@dubadub Ah, so Windows is taking the metadata to be part of the file extension and thus no preview occurs? Yes, that's not helpful.

If the spec was relaxed somewhat to allow leading spaces for metadata, might that work? e.g.

```
Chicken French.cook
Chicken French.jpg {0%step}
Chicken French.jpg {3%step}
```

I had a quick look at the Swift and Rust to see if I could ascertain the difficulty for this change but I don't know either language well enough to tell at a glance. Perhaps it would even be helpful to some people, e.g.

```
Chicken French.cook
Chicken French.jpg             {0%step}
Chicken French.jpg             {3%step}

Poke holes in @potato {2}.
```

Some might like to separate the "noise" of metadata from other things, and it might pay to be more relaxed and avoid annoying syntax errors for whitespace.

*However*, it's just an idea. I quite like it but I'm not averse to it being shot down 😅  If so, @deathau's suggestion might be a good middle ground.

Regards,
iain
